src/types/acls.rs

```diff
@@ -0,0 +1,387 @@
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`ListRights`] struct
+    pub fn list_rights(&self) -> &ListRights<'_> {
+        self.borrow_rights()
+    }
+}
+
+/// From [section 3.7 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.7).
+///
+/// Used by [`ListRightsResponse`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub struct ListRights<'a> {
+    /// The mailbox for the rights
+    pub(crate) mailbox: Cow<'a, str>,
+    /// The user identifier for the rights
+    pub(crate) identifier: Cow<'a, str>,
+    /// The set of rights that are always provided for this identifier
+    pub(crate) required: AclRights,
+    /// The set of rights that can be granted to the identifier
+    pub(crate) optional: AclRights,
+}
+
+impl ListRights<'_> {
+    /// Returns the mailbox for the rights
+    pub fn mailbox(&self) -> &str {
+        &*self.mailbox
+    }
+
+    /// Returns the user identifier for the rights
+    pub fn identifier(&self) -> &str {
+        &*self.identifier
+    }
+
+    /// Returns the set of rights that are always provided for this identifier
+    pub fn required(&self) -> &AclRights {
+        &self.required
+    }
+
+    /// Returns the set of rights that can be granted to the identifier
+    pub fn optional(&self) -> &AclRights {
+        &self.optional
+    }
+}
+
+/// From [section 3.8 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.8).
+///
+/// This is a wrapper around a single [`MyRights`].
+///
+/// The MYRIGHTS response from the [`Session::my_rights`] IMAP command
+#[self_referencing]
+pub struct MyRightsResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) rights: MyRights<'this>,
+}
+
+impl MyRightsResponse {
+    /// Parse the given input into a [`MyRights`] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        MyRightsResponseTryBuilder {
+            data: owned,
+            rights_builder: |input| {
+                // There should only be ONE single MYRIGHTS response
+                parse_until_done(input, unsolicited, |response| match response {
+                    Response::MyRights(a) => Ok(MapOrNot::Map(MyRights {
+                        mailbox: a.mailbox,
+                        rights: a.rights.into(),
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`MyRights`] struct
+    pub fn my_rights(&self) -> &MyRights<'_> {
+        self.borrow_rights()
+    }
+}
+
+/// From [section 3.8 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.8).
+///
+/// Used by [`MyRightsResponse`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
```

Comment:
This also does not need to be `#[non_exhaustive]` if it's only going to have private fields.